aws/logs_monitoring/enhanced_metrics.py

```diff
@@ -0,0 +1,445 @@
+        for enhanced_metric in enhanced_metrics:
+            enhanced_metric.submit_to_dd()
+    except Exception:
+        log.exception(
+            "Encountered an error while trying to parse and submit enhanced metrics"
+        )
+
+
+def generate_enhanced_lambda_metrics(logs, tags_cache):
+    """Parses Lambda logs for enhanced Lambda metrics and tags
+
+    Args:
+        logs (dict<str, multiple types>[]): the logs parsed from the event in the split method
+            Ex: [{
+                    "id": "34988208851106313984209006125707332605649155257376768001",
+                    "timestamp": 1568925546641,
+                    "message": "END RequestId: 2f676573-c16b-4207-993a-51fb960d73e2\\n",
+                    "aws": {
+                        "awslogs": {
+                            "logGroup": "/aws/lambda/function_log_generator",
+                            "logStream": "2019/09/19/[$LATEST]0225597e48f74a659916f0e482df5b92",
+                            "owner": "172597598159"
+                        },
+                        "function_version": "$LATEST",
+                        "invoked_function_arn": "arn:aws:lambda:us-east-1:172597598159:function:collect_logs_datadog_demo"
+                    },
+                    "lambda": {
+                        "arn": "arn:aws:lambda:us-east-1:172597598159:function:function_log_generator"
+                    },
+                    "ddsourcecategory": "aws",
+                    "ddtags": "env:demo,python_version:3.6,role:lambda,forwardername:collect_logs_datadog_demo,memorysize:128,forwarder_version:2.0.0,functionname:function_log_generator,env:none",
+                    "ddsource": "lambda",
+                    "service": "function_log_generator",
+                    "host": "arn:aws:lambda:us-east-1:172597598159:function:function_log_generator"
+                }]
+        tags_cache (LambdaTagsCache): used to apply the Lambda's custom tags to the metrics
+
+    Returns:
+        DatadogMetricPoint[], where each metric has all of its tags
+    """
+    enhanced_metrics = []
+    for log in logs:
+        log_function_arn = log.get("lambda", {}).get("arn")
+        log_message = log.get("message")
+        timestamp = log.get("timestamp")
+
+        # If the log dict is missing any of this data it's not a Lambda REPORT log and we move on
+        if not all(
+            (log_function_arn, log_message, timestamp, log_message.startswith("REPORT"))
+        ):
+            continue
+
+        parsed_metrics = parse_metrics_from_report_log(log_message)
+        if not parsed_metrics:
+            continue
+
+        # Add the tags from ARN, custom tags cache, and env var
+        tags_from_arn = parse_lambda_tags_from_arn(log_function_arn)
+        lambda_custom_tags = tags_cache.get_lambda_tags(log_function_arn)
+        # "ddtags" is the name of the key in the log event dict
+        tags_from_env_var = log.get("ddtags").split(",") if log.get("ddtags") else []
```

Comment:
I would not apply these tags for now, as the existing tags are mostly meant for logs, and I do not want user to lose tags in the future if we change our minds and generate these metrics from somewhere else. 

Also most of users can only have 4 tags, and submitting > 4 tags will completely ruin the metric. `lambda_custom_tags` is fine, since it's opt-in and we can make sure customers have distribution summaries enabled before enabling it.